### PR TITLE
Add example showing a TB clock generator in HDL

### LINKED ISSUE #2192: Add example showing the use of a clock generator implemented in HDL
We should have an example where a design is driven by both a clock generator implemented in HDL and in Python to show 
(a) the difference in simulation performance and 
(b) how the HDL version is done.

Related to #2115.

### PR DESCRIPTION
**Pull request by cmarqu**
Closes #2192

### CODE REVIEW

**Comment by garmin-mjames** [comment]
You need to add the new example in examples/Makefile.

File: examples/clkgen_in_hdl/test_clkgen.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+async def progress_reporter(tb_hdl):
+    """Report progress by printing intermediate counter values."""
+    while True:
+        # With the following Timer, we sleep exactly the right time
+        # to end up printing every 10000th counter value.
+        # We could also wake up on each counter value change and
+        # print the value if it was modulo 10000,
+        # but simulation performance would be much worse.
+        await Timer(10_000 * 10, "ns")
```
Comment (garmin-mjames) [inline examples/clkgen_in_hdl/test_clkgen.py:31]:
```suggestion
        await Timer(10000 * 10, "ns")
```
Underscores were apparently added in 3.6: https://docs.python.org/3.6/whatsnew/3.6.html#whatsnew36-pep515

Comment (ktbarrett) [reply examples/clkgen_in_hdl/test_clkgen.py:31]: Damn... they added a lot of QoL stuff in 3.6.

Comment (cmarqu) [reply examples/clkgen_in_hdl/test_clkgen.py]: Damn indeed... changed.

File: examples/clkgen_in_hdl/test_clkgen.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+# This file is public domain, it can be freely copied without restrictions.
+# SPDX-License-Identifier: CC0-1.0
+
+import cocotb
+from cocotb.clock import Clock
+from cocotb.triggers import Timer
+
+MAXCOUNT = 123456
```
Comment (ktbarrett) [inline examples/clkgen_in_hdl/test_clkgen.py:8]: This is a good case for a plusarg or environment variable communicating to cocotb.

File: documentation/source/examples.rst [hunk 3]
```diff
@@ -6,6 +6,44 @@ Apart from the examples covered with full tutorials in the previous sections,
...
+
+Here is the last part of the output of this example (with :envvar:`COCOTB_REDUCED_LOG_FMT` set),
+showing the different runtimes::
+
+    2469132.00ns INFO     ****************************************************************************************
+                          ** TEST                            PASS/FAIL  SIM TIME(NS)  REAL TIME(S)  RATIO(NS/S) **
+                          ****************************************************************************************
+                          ** test_clkgen.test_clkgen_hdl       PASS      1234566.00          0.03   41720542.34  **
+                          ** test_clkgen.test_clkgen_python    PASS      1234566.00         29.95     41219.68  **
+                          ****************************************************************************************
```
Comment (ktbarrett) [inline documentation/source/examples.rst:36]: Hmmmm... While it does show pretty effectively that using an HDL clock is a major improvement, I don't want to get their hopes up on a 100x performance increase across the board. Is there a way to temper their expectations?

Comment (garmin-mjames) [reply documentation/source/examples.rst:36]: Other things based on the clock need to happen. Right now it's either crossing into Python for every clock edge, or in the other case not at all until the end, so a sort of all-or-nothing.

Maybe change to a clock divider and use the slower clock in Python to count? That way a decent amount of "work" is happening in Python, but the speed-up will still be noticeable between the slower clock edges.

**Comment by cmarqu** [comment]
This is obsoleted by #89 once merged.

**Comment by cmarqu** [comment]
This will be obsoleted by the successor of #2293.

File: documentation/source/examples.rst (hunk 3 above)
Comment (jiao-x) [reply documentation/source/examples.rst:36]: Is there currently a way to speed up the simulation in this case when only python is used to generate the clock

**Pull request closed without merging**